File: src/MediaOrganizer.Android/ViewModels/MainViewModel.cs
using System.Collections.ObjectModel;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using MediaOrganizer.Android.Platforms;
using MediaOrganizer.Core;
using MediaOrganizer.Core.Models;
using MediaOrganizer.Shared.ViewModels;

namespace MediaOrganizer.Android.ViewModels;

/// <summary>抽屉导航项（FR-A7.1）：标签 + 页索引 + 失败文件角标（Badge=0 时角标隐藏）。
/// 可变 ObservableObject：角标变化时原地更新属性，避免替换实例破坏 ListBox 选中状态（P1-8）。</summary>
public sealed partial class NavItem : ObservableObject
{
    public NavItem(string label, int page)
    {
        Label = label;
        Page = page;
    }

    public string Label { get; }
    public int Page { get; }

    [ObservableProperty]
    private int _badge;

    public bool HasBadge => Badge > 0;

    partial void OnBadgeChanged(int value) => OnPropertyChanged(nameof(HasBadge));
}

/// <summary>
/// 应用外壳 VM（FR-A7）：顶部 AppBar 标题随页面切换、侧滑抽屉（整理工作台 / 失败文件 / 分析报告 / 设置）、
/// 底部状态栏聚合操作反馈（忙碌圆点 + 状态文本 + 上次分析摘要）。
/// 返回键行为（关抽屉/退出）由 MainActivity 的 OnBackPressedCallback 处理；
/// WakeLock 随分析/执行的忙碌状态自动持有/释放（NFR-A9）。
/// </summary>
public partial class MainViewModel : ViewModelBase
{
    public WorkbenchViewModel Workbench { get; }
    public FailedFilesViewModel FailedFiles { get; }
    public SettingsViewModel Settings { get; }
    public ReportViewModel Report { get; }

    private readonly WakeLockHolder _wakeLock;

    [ObservableProperty]
    private string _appTitle = "整理工作台";

    [ObservableProperty]
    private int _failedBadge;

    [ObservableProperty]
    private string _statusText = "就绪";

    [ObservableProperty]
    private string _statusDetail = "尚未分析";

    [ObservableProperty]
    private bool _isBusy;

    [ObservableProperty]
    private bool _drawerOpen;

    [ObservableProperty]
    private int _selectedPage; // 0 工作台 / 1 失败 / 2 报告 / 3 设置

    /// <summary>抽屉导航项（失败文件项带角标，角标变化时替换条目触发刷新）。</summary>
    public ObservableCollection<NavItem> NavItems { get; } =
    [
        new NavItem("整理工作台", 0),
        new NavItem("失败文件", 1) { Badge = 0 },
        new NavItem("分析报告", 2),
        new NavItem("设置", 3)
    ];

    public ViewModelBase? CurrentPage => SelectedPage switch
    {
        1 => FailedFiles,
        2 => Report,
        3 => Settings,
        _ => Workbench
    };

    public MainViewModel(
        WorkbenchViewModel workbench,
        FailedFilesViewModel failedFiles,
        SettingsViewModel settings,
        ReportViewModel report,
        WakeLockHolder wakeLock)
    {
        Workbench = workbench;
        FailedFiles = failedFiles;
        Settings = settings;
        Report = report;
        _wakeLock = wakeLock;
    }

    partial void OnSelectedPageChanged(int value)
    {
        AppTitle = value switch
        {
            1 => "失败文件",
            2 => "分析报告",
            3 => "设置",
            _ => "整理工作台"
        };
        OnPropertyChanged(nameof(CurrentPage));
    }

    [RelayCommand]
    private void ToggleDrawer() => DrawerOpen = !DrawerOpen;

    [RelayCommand]
    private void CloseDrawer() => DrawerOpen = false;

    [RelayCommand]
    private void Navigate(int page)
    {
        SelectedPage = page;
        DrawerOpen = false;
    }

    partial void OnFailedBadgeChanged(int value)
        => NavItems[1].Badge = value;

    public void RefreshFailedBadge(AnalysisResult? result)
    {
        FailedBadge = result?.Unparsed.Count ?? 0;
        StatusDetail = result is null
            ? "尚未分析"
            : $"上次分析: {result.Total:N0} 文件 · 成功 {result.Parsed.Count:N0}";
    }

    partial void OnIsBusyChanged(bool value)
    {
        if (value) _wakeLock.Acquire();
        else _wakeLock.Release();
    }

    /// <summary>
    /// 由 App 订阅 workbench.IsBusy 转发：忙碌时持 WakeLock（OnIsBusyChanged）。
    /// 注意：此处不可重置 StatusText——忙碌结束晚于结果/错误文本写入，会把「执行完成 / 执行失败：…」冲掉，
    /// 表现为操作"点了没反应"。状态栏文本统一由 workbench.StatusText 转发。
    /// </summary>
    public void SetBusy(bool busy) => IsBusy = busy;
}

File: src-flutter/lib/viewmodels/main_view_model.dart
/// 应用外壳 VM（对应 Android/ViewModels/MainViewModel.cs，FR-A7）：
/// AppBar 标题随页面切换、抽屉导航（整理工作台/失败文件/分析报告/设置）、
/// 底部状态栏聚合操作反馈（忙碌圆点 + 状态文本 + 上次分析摘要）。
library;

import 'package:flutter/foundation.dart';

import '../core/app_state.dart';
import '../core/format.dart';
import 'failed_files_view_model.dart';
import 'report_view_model.dart';
import 'settings_view_model.dart';
import 'workbench_view_model.dart';

/// 抽屉导航项：标签 + 页索引 + 失败文件角标（badge=0 时角标隐藏）。
class NavItem {
  final String label;
  final int page;
  final int badge;

  const NavItem(this.label, this.page, {this.badge = 0});

  bool get hasBadge => badge > 0;

  NavItem withBadge(int b) => NavItem(label, page, badge: b);
}

class MainViewModel extends ChangeNotifier {
  final AppState state;
  late final WorkbenchViewModel workbench;
  late final FailedFilesViewModel failedFiles;
  late final ReportViewModel report;
  late final SettingsViewModel settings;

  int selectedPage = 0; // 0 工作台 / 1 失败 / 2 报告 / 3 设置
  String appTitle = '整理工作台';
  int failedBadge = 0;
  String statusText = '就绪';
  String statusDetail = '尚未分析';
  bool isBusy = false;

  MainViewModel({required this.state, required String dataDir}) {
    workbench = WorkbenchViewModel(state: state, dataDir: dataDir);
    failedFiles = FailedFilesViewModel(state: state);
    report = ReportViewModel(dataDir: dataDir);
    settings = SettingsViewModel(state: state);

    // 工作台分析完成 → 刷新失败列表/角标/报告 + 上次分析摘要
    workbench.onAnalysisCompletedExternal = (result) {
      _refreshAfterAnalysis(result);
    };
    workbench.addListener(_mirrorWorkbenchStatus);

    // 配置/模式变更 → 工作台重建提取链；网络位置列表 → 刷新目标下拉
    state.addListener(() {
      workbench.rebuildChain();
      workbench.reloadOutputTargets();
    });
  }

  List<NavItem> get navItems => [
        const NavItem('整理工作台', 0),
        NavItem('失败文件', 1, badge: failedBadge),
        const NavItem('分析报告', 2),
        const NavItem('设置', 3),
      ];

  void navigate(int page) {
    if (selectedPage == page) return;
    selectedPage = page;
    appTitle = switch (page) {
      1 => '失败文件',
      2 => '分析报告',
      3 => '设置',
      _ => '整理工作台',
    };
    notifyListeners();
  }

  void _refreshAfterAnalysis(AnalysisResult result) {
    failedFiles.refresh(result);
    report.set(result, workbench.displayTarget);
    failedBadge = result.unparsed.length;
    statusDetail =
        '上次分析: ${formatN0(result.total)} 文件 · 成功 ${formatN0(result.parsed.length)}';
    notifyListeners();
  }

  /// 工作台状态镜像到底部状态栏。
  void _mirrorWorkbenchStatus() {
    statusText = workbench.statusText;
    final busy = workbench.isBusy;
    if (isBusy && !busy) {
      statusDetail = workbench.hasResult
          ? '上次分析: ${formatN0(workbench.totalFiles)} 文件 · 成功 ${formatN0(workbench.successFiles)}'
          : statusDetail;
    }
    isBusy = busy;
    notifyListeners();
  }

  @override
  void dispose() {
    workbench.removeListener(_mirrorWorkbenchStatus);
    super.dispose();
  }
}
